`java.lang.IndexOutOfBoundsException` occurred when previewing SQL with placeholder

## Bug Report

### Which version of ShardingSphere did you use?
master - cda8c9b4026b94e4ca2437ab00d8fd3325b82f99

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
Proxy

### Expected behavior
SQL preview result.

### Actual behavior

MySQL Proxy
```
[ERROR] 2021-10-19 17:34:12.268 [ShardingSphere-Command-0] o.a.s.p.f.c.CommandExecutorTask - Exception occur: 
java.lang.IndexOutOfBoundsException: Index: 0
	at java.util.Collections$EmptyList.get(Collections.java:4456)
	at org.apache.shardingsphere.sharding.route.engine.condition.generator.ConditionValue.getValue(ConditionValue.java:50)
	at org.apache.shardingsphere.sharding.route.engine.condition.generator.ConditionValue.getValue(ConditionValue.java:41)
	at org.apache.shardingsphere.sharding.route.engine.condition.generator.ConditionValue.<init>(ConditionValue.java:36)
	at org.apache.shardingsphere.sharding.route.engine.condition.generator.impl.ConditionValueCompareOperatorGenerator.generate(ConditionValueCompareOperatorGenerator.java:64)
	at org.apache.shardingsphere.sharding.route.engine.condition.generator.ConditionValueGeneratorFactory.generate(ConditionValueGeneratorFactory.java:51)
	at org.apache.shardingsphere.sharding.route.engine.condition.engine.impl.WhereClauseShardingConditionEngine.createShardingConditionValueMap(WhereClauseShardingConditionEngine.java:123)
	at org.apache.shardingsphere.sharding.route.engine.condition.engine.impl.WhereClauseShardingConditionEngine.createShardingConditions(WhereClauseShardingConditionEngine.java:85)
	at org.apache.shardingsphere.sharding.route.engine.condition.engine.impl.WhereClauseShardingConditionEngine.createShardingConditions(WhereClauseShardingConditionEngine.java:75)
	at org.apache.shardingsphere.sharding.route.engine.ShardingSQLRouter.createShardingConditions(ShardingSQLRouter.java:66)
	at org.apache.shardingsphere.sharding.route.engine.ShardingSQLRouter.createRouteContext(ShardingSQLRouter.java:50)
	at org.apache.shardingsphere.sharding.route.engine.ShardingSQLRouter.createRouteContext(ShardingSQLRouter.java:44)
	at org.apache.shardingsphere.infra.route.engine.impl.PartialSQLRouteExecutor.route(PartialSQLRouteExecutor.java:62)
	at org.apache.shardingsphere.infra.route.engine.SQLRouteEngine.route(SQLRouteEngine.java:52)
	at org.apache.shardingsphere.infra.context.kernel.KernelProcessor.route(KernelProcessor.java:54)
	at org.apache.shardingsphere.infra.context.kernel.KernelProcessor.generateExecutionContext(KernelProcessor.java:46)
	at org.apache.shardingsphere.proxy.backend.text.distsql.ral.advanced.PreviewDistSQLBackendHandler.execute(PreviewDistSQLBackendHandler.java:84)
	at org.apache.shardingsphere.proxy.frontend.mysql.command.query.text.query.MySQLComQueryPacketExecutor.execute(MySQLComQueryPacketExecutor.java:57)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:99)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:72)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```


PostgreSQL Proxy
```
[ERROR] 2021-10-19 17:35:16.747 [Connection-3-ThreadExecutor] o.a.s.p.f.c.CommandExecutorTask - Exception occur: 
java.lang.IndexOutOfBoundsException: Index: 0
	at java.util.Collections$EmptyList.get(Collections.java:4456)
	at org.apache.shardingsphere.sharding.route.engine.condition.generator.ConditionValue.getValue(ConditionValue.java:50)
	at org.apache.shardingsphere.sharding.route.engine.condition.generator.ConditionValue.getValue(ConditionValue.java:41)
	at org.apache.shardingsphere.sharding.route.engine.condition.generator.ConditionValue.<init>(ConditionValue.java:36)
	at org.apache.shardingsphere.sharding.route.engine.condition.generator.impl.ConditionValueCompareOperatorGenerator.generate(ConditionValueCompareOperatorGenerator.java:64)
	at org.apache.shardingsphere.sharding.route.engine.condition.generator.ConditionValueGeneratorFactory.generate(ConditionValueGeneratorFactory.java:51)
	at org.apache.shardingsphere.sharding.route.engine.condition.engine.impl.WhereClauseShardingConditionEngine.createShardingConditionValueMap(WhereClauseShardingConditionEngine.java:123)
	at org.apache.shardingsphere.sharding.route.engine.condition.engine.impl.WhereClauseShardingConditionEngine.createShardingConditions(WhereClauseShardingConditionEngine.java:85)
	at org.apache.shardingsphere.sharding.route.engine.condition.engine.impl.WhereClauseShardingConditionEngine.createShardingConditions(WhereClauseShardingConditionEngine.java:75)
	at org.apache.shardingsphere.sharding.route.engine.ShardingSQLRouter.createShardingConditions(ShardingSQLRouter.java:66)
	at org.apache.shardingsphere.sharding.route.engine.ShardingSQLRouter.createRouteContext(ShardingSQLRouter.java:50)
	at org.apache.shardingsphere.sharding.route.engine.ShardingSQLRouter.createRouteContext(ShardingSQLRouter.java:44)
	at org.apache.shardingsphere.infra.route.engine.impl.PartialSQLRouteExecutor.route(PartialSQLRouteExecutor.java:62)
	at org.apache.shardingsphere.infra.route.engine.SQLRouteEngine.route(SQLRouteEngine.java:52)
	at org.apache.shardingsphere.infra.context.kernel.KernelProcessor.route(KernelProcessor.java:54)
	at org.apache.shardingsphere.infra.context.kernel.KernelProcessor.generateExecutionContext(KernelProcessor.java:46)
	at org.apache.shardingsphere.proxy.backend.text.distsql.ral.advanced.PreviewDistSQLBackendHandler.execute(PreviewDistSQLBackendHandler.java:84)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.text.PostgreSQLComQueryExecutor.execute(PostgreSQLComQueryExecutor.java:71)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:99)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:72)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
Preview any SQL with placeholder.
```sql
preview select * from t where id = ?
```

--- a/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/route/engine/condition/generator/ConditionValue.java
+++ b/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/route/engine/condition/generator/ConditionValue.java
@@ -47,7 +47,9 @@ public final class ConditionValue {
     }
     
     private Comparable<?> getValue(final ParameterMarkerExpressionSegment expressionSegment, final List<Object> parameters) {
-        Object result = parameters.get(expressionSegment.getParameterMarkerIndex());
+        int parameterMarkerIndex = expressionSegment.getParameterMarkerIndex();
+        Preconditions.checkArgument(parameterMarkerIndex < parameters.size(), "Parameter marker expression must have corresponding parameter value.");
+        Object result = parameters.get(parameterMarkerIndex);
         Preconditions.checkArgument(result instanceof Comparable, "Sharding value must implements Comparable.");
         return (Comparable<?>) result;
     }

--- a/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/route/engine/condition/generator/impl/ConditionValueBetweenOperatorGeneratorTest.java
+++ b/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/route/engine/condition/generator/impl/ConditionValueBetweenOperatorGeneratorTest.java
@@ -17,23 +17,29 @@
 
 package org.apache.shardingsphere.sharding.route.engine.condition.generator.impl;
 
+import com.google.common.collect.Range;
 import org.apache.shardingsphere.infra.datetime.DatetimeService;
 import org.apache.shardingsphere.sharding.route.engine.condition.Column;
 import org.apache.shardingsphere.sharding.route.engine.condition.value.RangeShardingConditionValue;
 import org.apache.shardingsphere.sharding.route.engine.condition.value.ShardingConditionValue;
 import org.apache.shardingsphere.spi.ShardingSphereServiceLoader;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.column.ColumnSegment;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.BetweenExpression;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.ExpressionSegment;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.complex.CommonExpressionSegment;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.simple.LiteralExpressionSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.simple.ParameterMarkerExpressionSegment;
 import org.apache.shardingsphere.sql.parser.sql.common.util.SafeNumberOperationUtil;
+import org.apache.shardingsphere.sql.parser.sql.common.value.identifier.IdentifierValue;
 import org.junit.Test;
 
+import java.util.Arrays;
 import java.util.Calendar;
 import java.util.Date;
 import java.util.LinkedList;
 import java.util.Optional;
 
+import static org.hamcrest.CoreMatchers.instanceOf;
 import static org.hamcrest.CoreMatchers.is;
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
@@ -121,5 +127,30 @@ public final class ConditionValueBetweenOperatorGeneratorTest {
         assertThat(rangeShardingConditionValue.getColumnName(), is(column.getName()));
         assertThat(rangeShardingConditionValue.getTableName(), is(column.getTableName()));
         assertTrue(rangeShardingConditionValue.getValueRange().upperEndpoint().before(after));
+    }
+    
+    @SuppressWarnings("unchecked")
+    @Test
+    public void assertGenerateConditionValueWithParameter() {
+        ColumnSegment left = new ColumnSegment(0, 0, new IdentifierValue("id"));
+        ParameterMarkerExpressionSegment between = new ParameterMarkerExpressionSegment(0, 0, 0);
+        ParameterMarkerExpressionSegment and = new ParameterMarkerExpressionSegment(0, 0, 1);
+        BetweenExpression predicate = new BetweenExpression(0, 0, left, between, and, false);
+        Optional<ShardingConditionValue> actual = generator.generate(predicate, column, Arrays.asList(1, 2));
+        assertTrue(actual.isPresent());
+        assertThat(actual.get(), instanceOf(RangeShardingConditionValue.class));
+        RangeShardingConditionValue<Integer> conditionValue = (RangeShardingConditionValue<Integer>) actual.get();
+        assertThat(conditionValue.getTableName(), is("tbl"));
+        assertThat(conditionValue.getColumnName(), is("id"));
+        assertThat(conditionValue.getValueRange(), is(Range.closed(1, 2)));
+    }
+    
+    @Test(expected = IllegalArgumentException.class)
+    public void assertGenerateConditionValueWithoutParameter() {
+        ColumnSegment left = new ColumnSegment(0, 0, new IdentifierValue("id"));
+        ParameterMarkerExpressionSegment between = new ParameterMarkerExpressionSegment(0, 0, 0);
+        ParameterMarkerExpressionSegment and = new ParameterMarkerExpressionSegment(0, 0, 1);
+        BetweenExpression predicate = new BetweenExpression(0, 0, left, between, and, false);
+        generator.generate(predicate, column, new LinkedList<>());
     }
 }

--- a/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/route/engine/condition/generator/impl/ConditionValueCompareOperatorGeneratorTest.java
+++ b/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/route/engine/condition/generator/impl/ConditionValueCompareOperatorGeneratorTest.java
@@ -26,12 +26,18 @@ import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.column.Column
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.BinaryOperationExpression;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.complex.CommonExpressionSegment;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.simple.LiteralExpressionSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.simple.ParameterMarkerExpressionSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.value.identifier.IdentifierValue;
 import org.junit.Test;
 
+import java.util.Collections;
 import java.util.LinkedList;
 import java.util.Optional;
 
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.CoreMatchers.is;
 import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
 import static org.mockito.Mockito.mock;
 
@@ -106,5 +112,28 @@ public final class ConditionValueCompareOperatorGeneratorTest {
         Optional<ShardingConditionValue> shardingConditionValue = generator.generate(rightValue, column, new LinkedList<>());
         assertTrue(shardingConditionValue.isPresent());
         assertFalse(((ListShardingConditionValue<Integer>) shardingConditionValue.get()).getValues().isEmpty());
+    }
+    
+    @SuppressWarnings("unchecked")
+    @Test
+    public void assertGenerateConditionValueWithParameter() {
+        ColumnSegment left = new ColumnSegment(0, 0, new IdentifierValue("id"));
+        ParameterMarkerExpressionSegment right = new ParameterMarkerExpressionSegment(0, 0, 0);
+        BinaryOperationExpression predicate = new BinaryOperationExpression(0, 0, left, right, "=", "id = ?");
+        Optional<ShardingConditionValue> actual = generator.generate(predicate, column, Collections.singletonList(1));
+        assertTrue(actual.isPresent());
+        assertThat(actual.get(), instanceOf(ListShardingConditionValue.class));
+        ListShardingConditionValue<Integer> conditionValue = (ListShardingConditionValue<Integer>) actual.get();
+        assertThat(conditionValue.getTableName(), is("tbl"));
+        assertThat(conditionValue.getColumnName(), is("id"));
+        assertThat(conditionValue.getValues(), is(Collections.singletonList(1)));
+    }
+    
+    @Test(expected = IllegalArgumentException.class)
+    public void assertGenerateConditionValueWithoutParameter() {
+        ColumnSegment left = new ColumnSegment(0, 0, new IdentifierValue("order_id"));
+        ParameterMarkerExpressionSegment right = new ParameterMarkerExpressionSegment(0, 0, 0);
+        BinaryOperationExpression predicate = new BinaryOperationExpression(0, 0, left, right, "=", "order_id = ?");
+        generator.generate(predicate, column, new LinkedList<>());
     }
 }

--- a/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/route/engine/condition/generator/impl/ConditionValueInOperatorGeneratorTest.java
+++ b/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/route/engine/condition/generator/impl/ConditionValueInOperatorGeneratorTest.java
@@ -22,17 +22,22 @@ import org.apache.shardingsphere.sharding.route.engine.condition.Column;
 import org.apache.shardingsphere.sharding.route.engine.condition.value.ListShardingConditionValue;
 import org.apache.shardingsphere.sharding.route.engine.condition.value.ShardingConditionValue;
 import org.apache.shardingsphere.spi.ShardingSphereServiceLoader;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.column.ColumnSegment;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.InExpression;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.ListExpression;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.complex.CommonExpressionSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.simple.ParameterMarkerExpressionSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.value.identifier.IdentifierValue;
 import org.junit.Before;
 import org.junit.Test;
 
+import java.util.Collections;
 import java.util.Date;
 import java.util.LinkedList;
 import java.util.Optional;
 
 import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.CoreMatchers.is;
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
 
@@ -55,5 +60,30 @@ public final class ConditionValueInOperatorGeneratorTest {
         Optional<ShardingConditionValue> shardingConditionValue = generator.generate(inExpression, column, new LinkedList<>());
         assertTrue(shardingConditionValue.isPresent());
         assertThat(((ListShardingConditionValue) shardingConditionValue.get()).getValues().iterator().next(), instanceOf(Date.class));
+    }
+    
+    @SuppressWarnings("unchecked")
+    @Test
+    public void assertGenerateConditionValueWithParameter() {
+        ColumnSegment left = new ColumnSegment(0, 0, new IdentifierValue("id"));
+        ListExpression right = new ListExpression(0, 0);
+        right.getItems().add(new ParameterMarkerExpressionSegment(0, 0, 0));
+        InExpression predicate = new InExpression(0, 0, left, right, false);
+        Optional<ShardingConditionValue> actual = generator.generate(predicate, column, Collections.singletonList(1));
+        assertTrue(actual.isPresent());
+        assertThat(actual.get(), instanceOf(ListShardingConditionValue.class));
+        ListShardingConditionValue<Integer> conditionValue = (ListShardingConditionValue<Integer>) actual.get();
+        assertThat(conditionValue.getTableName(), is("tbl"));
+        assertThat(conditionValue.getColumnName(), is("id"));
+        assertThat(conditionValue.getValues(), is(Collections.singletonList(1)));
+    }
+    
+    @Test(expected = IllegalArgumentException.class)
+    public void assertGenerateConditionValueWithoutParameter() {
+        ColumnSegment left = new ColumnSegment(0, 0, new IdentifierValue("order_id"));
+        ListExpression right = new ListExpression(0, 0);
+        right.getItems().add(new ParameterMarkerExpressionSegment(0, 0, 0));
+        InExpression predicate = new InExpression(0, 0, left, right, false);
+        generator.generate(predicate, column, new LinkedList<>());
     }
 }

